src/main/java/org/opensearch/security/identity/SecurityTokenManager.java

```diff
@@ -0,0 +1,141 @@
+
+    private JwtVendor jwtVendor = null;
+    private ConfigModel configModel = null;
+
+    public SecurityTokenManager(final ClusterService cs, final ThreadPool threadPool, final UserService userService) {
+        this.cs = cs;
+        this.threadPool = threadPool;
+        this.userService = userService;
+    }
+
+    @Subscribe
+    public void onConfigModelChanged(final ConfigModel configModel) {
+        this.configModel = configModel;
+    }
+
+    @Subscribe
+    public void onDynamicConfigModelChanged(final DynamicConfigModel dcm) {
+        final Settings oboSettings = dcm.getDynamicOnBehalfOfSettings();
+        final Boolean enabled = oboSettings.getAsBoolean("enabled", false);
+        if (enabled) {
+            jwtVendor = createJwtVendor(oboSettings);
+        } else {
+            jwtVendor = null;
+        }
+    }
+
+    /** For testing */
+    JwtVendor createJwtVendor(final Settings settings) {
+        try {
+            return new JwtVendor(settings, Optional.empty());
+        } catch (final Exception ex) {
+            logger.error("Unable to create the JwtVendor instance", ex);
+            return null;
+        }
+    }
+
+    public boolean issueOnBehalfOfTokenAllowed() {
+        return jwtVendor != null && configModel != null;
+    }
+
+    @Override
+    public ExpiringBearerAuthToken issueOnBehalfOfToken(final Subject subject, final OnBehalfOfClaims claims) {
+        if (!issueOnBehalfOfTokenAllowed()) {
+            // TODO: link that doc!
+            throw new OpenSearchSecurityException(
+                "The OnBehalfOf token generation is not enabled, see {link to doc} for more information on this feature."
+            );
+        }
+
+        if (subject != null && !(subject instanceof NoopSubject)) {
+            logger.warn("Unsupported subject for OnBehalfOfToken token generation, {}", subject);
+            throw new IllegalArgumentException("Unsupported subject to generate OnBehalfOfToken");
+        }
+
+        if (Strings.isNullOrEmpty(claims.getAudience())) {
+            throw new IllegalArgumentException("Claims must be supplied with an audience value");
+        }
+
+        final User user = threadPool.getThreadContext().getTransient(ConfigConstants.OPENDISTRO_SECURITY_USER);
+        if (user == null) {
+            throw new OpenSearchSecurityException("Unsupported user to generate OnBehalfOfToken");
+        }
+
+        final TransportAddress callerAddress = null; /* OBO tokens must not roles based on location from network address */
+        final Set<String> mappedRoles = configModel.mapSecurityRoles(user, callerAddress);
+
+        try {
+            return jwtVendor.createJwt(
+                cs.getClusterName().value(),
+                user.getName(),
+                claims.getAudience(),
+                claims.getExpiration(),
+                mappedRoles.stream().collect(Collectors.toList()),
+                user.getRoles().stream().collect(Collectors.toList()),
```

Comment:
Can this lookup the audience (extension)'s setting to see if this token should include the backend roles in the payload?

Reply:
Lets look into this later as a separate change.